### PR TITLE
fix: do not freeze process.env in dev

### LINKED ISSUE #12618: Unable to resolve import.meta.env.xxx
### Astro Info

```block
PS F:\eya46eya46> astro info
Astro                    v5.0.2
Node                     v22.12.0
System                   Windows (x64)
Package Manager          unknown // pnpm
Output                   server
Adapter                  @astrojs/node
Integrations             @astrojs/tailwind
```

### If this issue only occurs in one browser, which browser is a problem?

_No response_

### Describe the Bug

my code [src/utils/wakatime.ts](https://github.com/eya46/eya46eya46/blob/master/src/utils/wakatime.ts#L6)

> The following content is from a translation.

Reproduce the bug(?)
```js
// astro.config.mjs
export default defineConfig({
  output: "server",
  adapter: node({
    mode: "standalone",
  }),
}

// any where
import.meta.env.xxx
```
It works correctly when using `astro dev`, but after building, it doesn't work properly when running `astro preview` or `node dist/server/entry.mjs`.

---

In the new version, the source code `const WAKATIME_TOKEN = import.meta.env.WAKATIME_TOKEN;` is compiled into `const WAKATIME_TOKEN = process.env.WAKATIME_TOKEN;`.

Previously, `import.meta.env.WAKATIME_TOKEN` would be directly replaced with the actual value. However, now it is replaced with `process.env.xxx`, making it unusable.

The .env file in the root directory is properly configured.
The error message is as follows:
```js
// dist/server/pages/index.astro.mjs
const WAKATIME_TOKEN = process.env.WAKATIME_TOKEN;
if (!WAKATIME_TOKEN) throw new Error("WAKATIME_TOKEN is not defined");
```

```
15:19:11 [ERROR] Error: WAKATIME_TOKEN is not defined
    at file:///F:/eya46eya46/dist/server/pages/index.astro.mjs:121:28
    at ModuleJob.run (node:internal/modules/esm/module_job:271:25)
    at async onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:547:26)
    at async AppPipeline.getModuleForRoute (file:///F:/eya46eya46/dist/server/chunks/_@astrojs-ssr-adapter_Boj1eyKp.mjs:2267:16)
    at async NodeApp.render (file:///F:/eya46eya46/dist/server/chunks/_@astrojs-ssr-adapter_Boj1eyKp.mjs:2447:19)
    at async file:///F:/eya46eya46/dist/server/chunks/_@astrojs-ssr-adapter_Boj1eyKp.mjs:2806:30
```

### What's the expected result?

Able to correctly handle `import.meta.env.xxx` or `.env` files.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/withastro-astro-drnr9z?file=src%2Fpages%2Findex.astro

### Participation

- [ ] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by florian-lefebvre**
## Changes

- In dev, inline env in the virtual module
- Cleans up unused stuff
- Closes PLT-2684
- Closes #12618

## Testing

- [x] Tests should still pass
- [x] Test manually
- [x] Test on a private project 

## Docs

N/A

### CODE REVIEW

File: packages/astro/src/env/vite-plugin-env.ts [hunk 1]
```diff
@@ -11,26 +11,41 @@ import {
...
 			const loadedEnv = loadEnv(mode, fileURLToPath(settings.config.root), '');
-			for (const [key, value] of Object.entries(loadedEnv)) {
-				if (value !== undefined) {
-					process.env[key] = value;
+
+			if (isDev) {
+				if (!(ENV_SYMBOL in globalThis)) {
+					setGetEnv((key) => (globalThis as any)[ENV_SYMBOL][key]);
+				}
+				(globalThis as any)[ENV_SYMBOL] = loadedEnv;
```
Comment (florian-lefebvre) [inline packages/astro/src/env/vite-plugin-env.ts:43]: That's the actual fix

File: packages/astro/templates/env.mjs [hunk 2]
```diff
@@ -2,14 +2,19 @@
...
-	getEnv,
+	getEnv as _getEnv,
 	getEnvFieldType,
 	setOnSetGetEnv,
 	validateEnvVariable,
 } from 'astro/env/runtime';
 
+const getEnv = (key) => {
+	// @@GET_ENV@@
+	return _getEnv(key);
```
Comment (ematipico) [inline packages/astro/templates/env.mjs:13]: This code will end up with two `return` statements, which isn't the best. Maybe we can leave the placeholder and add an `else` clause to the `if` we have `vite-plugin-env.ts`, line 157, which will replace the placeholder with `return _getEnv(key)`

**Review by ematipico — APPROVED** [review]
Looks good! I left a nit around the template

**Review by bluwy — CHANGES_REQUESTED** [review]
Blocking until we have a solution for https://github.com/withastro/astro/issues/12618

**Comment by florian-lefebvre** [comment]
@bluwy I think I have a fix for that issue, see [`4da93c6` (#12585)](https://github.com/withastro/astro/pull/12585/commits/4da93c62f1a17806e4faffbbae1bbc9bd5e62738)

File: packages/astro/templates/env.mjs [hunk 3]
```diff
@@ -2,14 +2,23 @@
...
 import {
 	createInvalidVariablesError,
-	getEnv,
+	getEnv as _getEnv,
 	getEnvFieldType,
 	setOnSetGetEnv,
 	validateEnvVariable,
 } from 'astro/env/runtime';
 
+// @ts-ignore
```
Comment (ematipico) [inline packages/astro/templates/env.mjs:11]:
```suggestion
// @ts-expect-error
```

**Review by ematipico — APPROVED** [review]
The code looks good. I can help you to test it with a preview release

**Comment by bluwy** [comment]
With the latest fix, I think it would fix the regression for now. But it doesn't fix the case where `.env` file updates are handled I think. If so, I'm still fine with merging this as a stop gap though.

**Comment by florian-lefebvre** [comment]
I'll test it on monday

**Comment by florian-lefebvre** [comment]
@bluwy I just tested manually in the basics example woth `import.meta.env`, `process.env` and `astro:env/server` and dev reloading works fine

**Pull request merged**